src/debugger/android/androidPlatform.ts

```diff
@@ -19,13 +20,56 @@ export class AndroidPlatform implements IAppPlatform {
 
     public enableJSDebuggingMode(runOptions: IRunOptions): Q.Promise<void> {
         let pkg = new Package(runOptions.projectRoot);
-
         return pkg.name()
             .then(appName => new PackageNameResolver(appName).resolvePackageName(runOptions.projectRoot))
             .then(packageName => {
-                let enableDebugCommand = `adb shell am broadcast -a "${packageName.toLowerCase()}.RELOAD_APP_ACTION" --ez jsproxy true`;
+                let deviceResolver = new DeviceResolver();
                 let cexec = new CommandExecutor(runOptions.projectRoot);
-                return cexec.execute(enableDebugCommand);
+                let debugTarget: string = null;
+                return deviceResolver.getConnectedDevices()
+                    .then((devices: IDevice[]) => {
+                        if (devices.length > 1) {
+                            /* more than one device or emulator */
+                            debugTarget = this.getTargetEmulator(runOptions, devices);
+                        }
+                    })
+                    .then(() => {
+                        if (debugTarget) {
+                            /* Launching is needed only if we have more than one device available */
+                            let launchAppCommand = `adb -s ${debugTarget} shell am start -n com.awesomeproject/.MainActivity`;
+                            return cexec.execute(launchAppCommand);
+                        }
+                    })
+                    .then(() => {
+                        let enableDebugCommand = `adb ${debugTarget ? "-s " + debugTarget : ""} shell am broadcast -a "${packageName.toLowerCase()}.RELOAD_APP_ACTION" --ez jsproxy true`;
+                        return cexec.execute(enableDebugCommand);
+                    });
             });
     }
+
+    /**
+     * Returns the target emulator, using the following logic:
+     * *  If an emulator is specified and it is connected, use that one.
+     * *  Otherwise, use the first one in the list.
+     */
+    private getTargetEmulator(runOptions: IRunOptions, devices: IDevice[]): string {
+        let activeFilterFunction = (device: IDevice) => {
+            return device.status === "device";
+        };
+
+        let targetFilterFunction = (device: IDevice) => {
+            return device.id === runOptions.target && activeFilterFunction(device);
+        };
+
+        if (runOptions.target && devices) {
+            /* check if the specified target is active */
+            if (devices.filter(targetFilterFunction).length > 0) {
+                return runOptions.target;
+            }
+        }
+
+        /* return the first active device in the list */
+        let activeDevices = devices && devices.filter(activeFilterFunction);
+        return activeDevices && activeDevices[0] && activeDevices[0].id;
```

Comment:
deviceResolver.getConnectedDevices() never returns null, and this is a private method, so I think it's okay to add a "non null parameters here" precondition.

If we want to be prepared in case the code changes, I'd add an assert.notNull(devices) at the top of the method, explicitly verifying the precondition.

If we stop using preconditions and assumptions on our private methods, then we'll have to add null checks in every private method before using an argument which will be a nightmare.